### PR TITLE
next -> redux opinionated

### PR DESCRIPTION
**Pull request by sergiokopplin**
Might need some help here.

### CODE REVIEW

**Comment by ctrlplusb** [comment]
Awesome man.  You are a legend for picking this up.  I'll review within the next few days. 💖

**Comment by sergiokopplin** [comment]
Hey @ctrlplusb, i couldn't find the problem to solve the bug.

Do you know if is related with react-jobs or with react-async-component?
I want to solve, but i need the first step, there is a lot of thing going there!

**Comment by birkir** [comment]
I got react-jobs and react-async-components working in a mobx environment if you want to take a look: https://github.com/ueno-llc/starter-kit-universally/pull/1/files

**Comment by ctrlplusb** [comment]
FYI, some BIG updates coming to both `react-async-component` and `react-jobs`.  I have a new interop layer that allows both of them to share the same tree walk parse, therefore only requiring a single parse. And it will resolve any number of nested async/job components. 🎉

**Comment by sergiokopplin** [comment]
Sorry guys, vacation time.
Waiting the merge. o//

**Comment by ctrlplusb** [comment]
Hey all, 

react-async-component upgraded.

Next up react-jobs - a big rewrite that will impact this PR in a big manner.

@sergiokopplin if you haven't already can you add me as being able to contribute to your PR. Thanks

**Comment by sergiokopplin** [comment]
@ctrlplusb added. Thanks for the help!

**Comment by sergiokopplin** [comment]
@ctrlplusb hey man, i'm back from vacations, i have some time to help here!

---

We're having some problems here.
The application must wait the request return to render the pages, but it's returning before the success. So, if you disable the JS in the browser, it will not work.

I still have the bug (might be a bug) in my application. Can you help with that?
I want to help in the merge, so i can upgrade my application here and also solve the bug.

Thanks again man. Awesome work here!

File: server/middleware/reactApplication/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  runJobs(app).then(({ appWithAsyncComponents, state, STATE_IDENTIFIER }) => {
+    // Generate the html response.
+    const html = renderToStaticMarkup(
+      <ServerHTML
+        reactAppString={renderToString(appWithAsyncComponents)}
+        nonce={nonce}
+        helmet={Helmet.rewind()}
+        asyncComponents={{ state, STATE_IDENTIFIER }}
+        initialState={getState()}
+        jobsTate={{ state, STATE_IDENTIFIER }}
```
Comment (sergiokopplin) [inline server/middleware/reactApplication/index.js]: @Neophy7e hey man, i'm not sure about this details, i need some help from @ctrlplusb to understand this.

**Comment by sergiokopplin** [comment]
@ctrlplusb i tried to merge the fork/redux-opinionated with the upstream/next but it has some conflicts that i'm not sure how to solve.

**Comment by ctrlplusb** [comment]
No worries.  Good to have you back! :)

I'll make some time to do a big merge / update of that branch with the latest.  We will work against master now, as I put next into master - officially released v13.

I'll get back to you early next week. 👍

**Comment by sergiokopplin** [comment]
Thanks man!

**Comment by birkir** [comment]
I have it up and running, checkout the changes: https://github.com/ueno-llc/starter-kit-universally/commit/09af801006f21ad7e91a675ce8c4d44b02ea2ef7

**Comment by sergiokopplin** [comment]
@birkir thanks man!

**Comment by sergiokopplin** [comment]
@ctrlplusb i've solved some conflicts on the repo with the next/merge in pair with @birkir.
I'll that a look on some errors that is coming in my terminal tomorrow!

**Comment by sergiokopplin** [comment]
image

**Comment by sergiokopplin** [comment]
image

**Comment by ctrlplusb** [comment]
@sergiokopplin sorry I haven't been able to give you time on this.  I have a client project I am pushing hard on at the moment.  Time will come.  Looks like you are doing a great job thusfar!

**Comment by oyeanuj** [comment]
> Time will come

Hey @ctrlplusb! Just wanted to get a sense from you for when do you see this being merged and what parts need to happen before the merge? Looking forward to using it for a project of mine 😄

**Comment by sergiokopplin** [comment]
@oyeanuj it is almost ready! I had some troubles to construct the async pages, but the configurations seems to be ready!

File: shared/components/DemoApp/AsyncPostsRoute/Post/Post.js [hunk 2]
```diff
@@ -69,12 +67,23 @@ function Post() {
...
+//     //     prevProps.params.id !== nextProps.params.id,
+//     // },
+//   }),
 // )(Post);
 
-export default Post;
+export default withJob({
+  work: (props) => {
+    const { id } = props.match.params;
+    return fetch(`http://jsonplaceholder.typicode.com/posts/${id}`).then(r => r.json());
```
Comment (Vincz) [inline shared/components/DemoApp/AsyncPostsRoute/Post/Post.js]: The fetch function is only defined in the browser causing the job to failed on the server side. What I don't understand is why there is nothing in the server console saying the function doesn't exist.

Comment (sergiokopplin) [inline shared/components/DemoApp/AsyncPostsRoute/Post/Post.js]: @Vincz yeah, you're right!! I'm missing something, but don't now how to solve.

Comment (strues) [inline shared/components/DemoApp/AsyncPostsRoute/Post/Post.js]: Any luck with this? I just pulled down your branch to look into it

Comment (sergiokopplin) [inline shared/components/DemoApp/AsyncPostsRoute/Post/Post.js]: Not yet! Not working properly! The redux functions are not plugged into the code. The server side rendering is not rendering 100%. Im working based on @birkir fork!

**Comment by strues** [comment]
I was starting to dig in but had to drop my MacBook off at Apple for the second time in a month.

File: shared/components/DemoApp/AsyncPostsRoute/Post/Post.js [hunk 3]
```diff
@@ -78,12 +78,14 @@ function Post({ jobResult }) {
 export default withJob({
   work: (props) => {
     const { id } = props.match.params;
-    return fetch(`http://jsonplaceholder.typicode.com/posts/${id}`).then(r => r.json());
+    return fetch(`http://jsonplaceholder.typicode.com/posts/${id}`)
```
Comment (Vincz) [inline shared/components/DemoApp/AsyncPostsRoute/Post/Post.js]: The problem is still here. fetch function is undefined server side. You should use axios to perform the query as it works on both side. If you do, then the SSR works just fine.

**Comment by Vincz** [comment]
Hi guys !
I have detected two problem:
- Still using "fetch" in the Post component. Fetch doesn't work server side. You should use Axios (working on both sides) to perform the json query.
- Client side, the redux store is populated from window.__APP_STATE__ which is great as we want the store to be prepopulated by the server. But, the server side counterparts is missing. In this file, we should get the store state https://github.com/sergiokopplin/react-universally/blob/feature/redux-opinionated/server/middleware/reactApplication/index.js then add it to the window in https://github.com/sergiokopplin/react-universally/blob/feature/redux-opinionated/server/middleware/reactApplication/ServerHTML.js

**Comment by sergiokopplin** [comment]
Awesome work @Vincz. The axios thing is new for me, great catch! I'll update the rep! Thanks again

**Comment by Vincz** [comment]
I'd like to send you the patch but I have no idea how to add commits to someone else PR (I'm a git newbie).

So, for the axios part, it would be:

    return axios(`http://jsonplaceholder.typicode.com/posts/${id}`)
      .then(r => r.data)
      .catch(err => console.error(err));

And for the redux state, 
in `server/middleware/reactApplication/index.js`

    <ServerHTML
        reactAppString={appString}
        nonce={nonce}
        helmet={Helmet.rewind()}
        storeState={store.getState()}
        jobsState={jobContext.getState()}
        asyncComponentsState={asyncContext.getState()}
    />

and in `server/middleware/reactApplication/ServerHTML.js`

line 43:

    const { storeState, jobsState, asyncComponentsState, helmet, nonce, reactAppString } = props;

line 64 (add):

    // Bind our redux store state so the client knows how to hydrate his one
    ifElse(storeState)(() => inlineScript(`window.__APP_STATE__=${serialize(storeState)};`)),

**Comment by sergiokopplin** [comment]
@Vincz awesome!! thanks!
tomorrow i'll update the fork!

**Comment by brianfeister** [comment]
Not sure if this is of any interest to you guys, but I grabbed the fork from this PR and added `material-ui` which was a bit difficult for me, being my second day on the job as a React developer. No idea if this aligns with your goals in maintaining branches with other ingredients in the recipe.

Went down a rabbit hole with the `material-ui` folks due to the complexity in the redux-opinionated fork:
https://github.com/nathanmarks/jss-theme-reactor/issues/41#issuecomment-298260293

https://github.com/brianfeister/react-universally/commit/2e80248a2c46b68fe38fa7c022dad33fd07a84de

File: shared/components/DemoApp/AsyncPostsRoute/Post/Post.js (hunk 2 above)
Comment (birkir) [inline shared/components/DemoApp/AsyncPostsRoute/Post/Post.js]: Use axios for fetching data, its cancellable and also lightweight.

**Comment by sergiokopplin** [comment]
hey @brianfeister, we should talk with @ctrlplusb first about your ideas. :)

**Comment by sergiokopplin** [comment]
a73f171 fixed!
to test if it's working, disable the javascript on browser then open the post/route. The async functions are working properly fetching all the data on the server!

**Comment by ctrlplusb** [comment]
Sorry I have been a bit silent. Really warms my heart to see so many people pulling together on this one. Hope to help soon. 😘

**Comment by strues** [comment]
@sergiokopplin submitted a PR under your fork with a small oversight i saw this AM.

**Comment by sergiokopplin** [comment]
Thanks @strues

**Comment by strues** [comment]
@sergiokopplin Another PR submitted on your branch. The old PropTypes from react were being used on the post component. I went ahead and added a few more tests as well.

**Comment by sergiokopplin** [comment]
Thanks @strues!!!!! Awesome work!

**Comment by sergiokopplin** [comment]
@ctrlplusb i'm not sure if is ready to merge, can you take a look? Thanks

**Comment by strues** [comment]
It's certainly close to ready, if not completely ready. Havent noticed any other issues myself.

**Comment by brianfeister** [comment]
I think I uncovered a minor bug where wrapping components differently than you're doing in the boilerplate causes the async `/post/:id` route to choke, I think this might be a better approach

https://github.com/sergiokopplin/react-universally/pull/3

**Comment by sergiokopplin** [comment]
Nice catch @brianfeister !!!

**Comment by sergiokopplin** [comment]
Hey guys!
The branch has 1 problem which affects the render speed.

image

I was looking as seems to be a problem with the async wrapper.
Can someone helps me with that? Not finding the problem.

**Comment by sergiokopplin** [comment]
Fixed!

I changed this line from client/index:

```js
const rehydrateState = window.__JOBS_REHYDRATE_STATE__;
```

to:

```js
const rehydrateState = window.__JOBS_STATE__;
```

File: client/index.js [hunk 4]
```diff
@@ -30,7 +30,7 @@ const supportsHistory = 'pushState' in window.history;
 
 // Get any rehydrateState for the async components.
 // eslint-disable-next-line no-underscore-dangle
-const asyncComponentsRehydrateState = window.__ASYNC_COMPONENTS_REHYDRATE_STATE__;
+const asyncComponentsRehydrateState = window.__JOBS_STATE__;
```
Comment (maryokhin) [inline client/index.js]: Should this part in [server/middleware/reactApplication/ServerHTML.js](https://github.com/ctrlplusb/react-universally/blob/master/server/middleware/reactApplication/ServerHTML.js#L69) also reflect this?
```
 ifElse(asyncComponentsState)(() =>
      inlineScript(
        `window.__ASYNC_COMPONENTS_REHYDRATE_STATE__=${serialize(asyncComponentsState)};`,
      ),
    ),
```
also reflect this change?

**Comment by maryokhin** [comment]
Running this branch as-is gives me a bunch of errors in the console, is that normal?

screenshot 2017-05-31 13 48 06

**Comment by maryokhin** [comment]
@sergiokopplin yep, console is clean now.

**Comment by maryokhin** [comment]
I tried writing my own stack of actions, reducers and components following the pattern in this branch...but completely failed 😞 maybe someone can explain why there is no `fetchPost` in my props? 
```javascript
const mapDispatchToProps = {
  fetch: fetchPlaylists,
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withJob({
    work: (props) => {
      console.log("work:", props);
      const { playlists, fetch } = props;
      console.log("fetch:", fetch);

      if (playlists) {
        return true;
      }

      return fetch();
    },
  }),
)(PlaylistList);
```
screenshot 2017-06-01 00 05 39

**Comment by oyeanuj** [comment]
@sergiokopplin @ctrlplusb Ready to merge? 🎉

**Comment by strues** [comment]
LGTM

**Comment by ctrlplusb** [comment]
If @strues is happy lets go! 👍

**Comment by ctrlplusb** [comment]
Well done to @sergiokopplin for all his hard work on this one, and a huge thanks to everyone else who helped along the way. 👍

**Comment by sergiokopplin** [comment]
Awesome!!!! :)

**Comment by sergiokopplin** [comment]
@ctrlplusb can i add my name on the contributors :p?

**Comment by ctrlplusb** [comment]
Of course yes!  

Sorry I have been completely dropping the ball on managing this project at the moment.  I have so many other gears turning, with some exciting new developments.  I appreciate to the moon your contribution.  In fact, expect an invite. 👍

**Comment by sergiokopplin** [comment]
Thanks @ctrlplusb!

**Comment by sergiokopplin** [comment]
@ctrlplusb hey man! i've pushed a commit that was supposed to be a pr on the next branch to add me as contributor. sorry for that, my mistake! Can you add me on readme also? thanks

**Comment by ctrlplusb** [comment]
No worries, my sincere apologies for not having you in there already!

**Pull request merged**